File: python/code_provenance/registry.py
import requests
from code_provenance.models import ImageRef


def get_registry_token(registry: str, repo_path: str) -> str | None:
    """Get an anonymous pull token from an OCI registry."""
    if registry == "ghcr.io":
        url = "https://ghcr.io/token"
        params = {"scope": f"repository:{repo_path}:pull"}
    elif registry == "docker.io":
        url = "https://auth.docker.io/token"
        params = {
            "service": "registry.docker.io",
            "scope": f"repository:{repo_path}:pull",
        }
    else:
        return None

    try:
        resp = requests.get(url, params=params, timeout=10)
        resp.raise_for_status()
        return resp.json()["token"]
    except (requests.RequestException, KeyError):
        return None


def _registry_base_url(registry: str) -> str:
    if registry == "docker.io":
        return "https://registry-1.docker.io"
    return f"https://{registry}"


_MANIFEST_ACCEPT = ", ".join([
    "application/vnd.docker.distribution.manifest.v2+json",
    "application/vnd.oci.image.manifest.v1+json",
    "application/vnd.docker.distribution.manifest.list.v2+json",
    "application/vnd.oci.image.index.v1+json",
])

_INDEX_MEDIA_TYPES = {
    "application/vnd.docker.distribution.manifest.list.v2+json",
    "application/vnd.oci.image.index.v1+json",
}


def _resolve_manifest_to_config_digest(
    base_url: str, repo_path: str, reference: str, token: str, _depth: int = 0,
) -> str | None:
    """Resolve a manifest reference to a config blob digest, handling multi-arch indexes."""
    headers = {"Authorization": f"Bearer {token}", "Accept": _MANIFEST_ACCEPT}

    try:
        resp = requests.get(
            f"{base_url}/v2/{repo_path}/manifests/{reference}",
            headers=headers,
            timeout=10,
        )
        if resp.status_code != 200:
            return None
        data = resp.json()
    except (requests.RequestException, KeyError, ValueError):
        return None

    media_type = data.get("mediaType", "")

    # If it's an index/manifest list, pick the first amd64/linux manifest
    if media_type in _INDEX_MEDIA_TYPES:
        manifests = data.get("manifests", [])
        platform_digest = None
        for m in manifests:
            platform = m.get("platform", {})
            # Skip attestation manifests
            if platform.get("os") == "unknown":
                continue
            if platform.get("architecture") == "amd64" and platform.get("os") == "linux":
                platform_digest = m["digest"]
                break
        if not platform_digest and manifests:
            # Fall back to first non-attestation manifest
            for m in manifests:
                if m.get("platform", {}).get("os") != "unknown":
                    platform_digest = m["digest"]
                    break
        if not platform_digest:
            return None
        if _depth >= 3:
            return None
        # Recursively resolve the platform-specific manifest
        return _resolve_manifest_to_config_digest(base_url, repo_path, platform_digest, token, _depth + 1)

    # Single manifest — extract config digest
    return data.get("config", {}).get("digest")


def fetch_oci_labels(ref: ImageRef) -> dict[str, str]:
    """Fetch OCI labels from an image's config blob without pulling the image."""
    repo_path = f"{ref.namespace}/{ref.name}"
    token = get_registry_token(ref.registry, repo_path)
    if not token:
        return {}

    base_url = _registry_base_url(ref.registry)
    config_digest = _resolve_manifest_to_config_digest(base_url, repo_path, ref.tag, token)
    if not config_digest:
        return {}

    # Fetch config blob to read labels
    try:
        config_headers = {
            "Authorization": f"Bearer {token}",
            "Accept": "application/vnd.docker.container.image.v1+json",
        }
        config_resp = requests.get(
            f"{base_url}/v2/{repo_path}/blobs/{config_digest}",
            headers=config_headers,
            timeout=10,
            allow_redirects=True,
        )
        if config_resp.status_code != 200:
            return {}
        return config_resp.json().get("config", {}).get("Labels", {}) or {}
    except (requests.RequestException, KeyError):
        return {}

File: node/src/registry.ts
import type { ImageRef } from "./types.js";

/**
 * Get an anonymous pull token from an OCI registry.
 */
export async function getRegistryToken(
  registry: string,
  repoPath: string
): Promise<string | null> {
  let url: string;

  if (registry === "ghcr.io") {
    url = `https://ghcr.io/token?scope=repository:${encodeURIComponent(repoPath)}:pull`;
  } else if (registry === "docker.io") {
    url =
      `https://auth.docker.io/token?service=registry.docker.io` +
      `&scope=repository:${encodeURIComponent(repoPath)}:pull`;
  } else {
    return null;
  }

  try {
    const resp = await fetch(url, { signal: AbortSignal.timeout(10000) });
    if (!resp.ok) return null;
    const data = (await resp.json()) as { token?: string };
    return data.token ?? null;
  } catch {
    return null;
  }
}

function registryBaseUrl(registry: string): string {
  if (registry === "docker.io") return "https://registry-1.docker.io";
  return `https://${registry}`;
}

const MANIFEST_ACCEPT = [
  "application/vnd.docker.distribution.manifest.v2+json",
  "application/vnd.oci.image.manifest.v1+json",
  "application/vnd.docker.distribution.manifest.list.v2+json",
  "application/vnd.oci.image.index.v1+json",
].join(", ");

const INDEX_MEDIA_TYPES = new Set([
  "application/vnd.docker.distribution.manifest.list.v2+json",
  "application/vnd.oci.image.index.v1+json",
]);

interface ManifestPlatform {
  os?: string;
  architecture?: string;
}

interface ManifestEntry {
  digest: string;
  platform?: ManifestPlatform;
}

/**
 * Resolve a manifest reference to a config blob digest, handling multi-arch indexes.
 */
async function resolveManifestToConfigDigest(
  baseUrl: string,
  repoPath: string,
  reference: string,
  token: string,
  depth: number = 0
): Promise<string | null> {
  try {
    const resp = await fetch(
      `${baseUrl}/v2/${repoPath}/manifests/${reference}`,
      {
        headers: {
          Authorization: `Bearer ${token}`,
          Accept: MANIFEST_ACCEPT,
        },
        signal: AbortSignal.timeout(10000),
      }
    );
    if (!resp.ok) return null;
    const data = await resp.json();
    const mediaType: string = data.mediaType ?? "";

    // If it's an index/manifest list, pick amd64/linux
    if (INDEX_MEDIA_TYPES.has(mediaType)) {
      const manifests: ManifestEntry[] = data.manifests ?? [];
      let platformDigest: string | null = null;

      // Try amd64/linux first
      for (const m of manifests) {
        const platform = m.platform ?? {};
        if (platform.os === "unknown") continue;
        if (platform.architecture === "amd64" && platform.os === "linux") {
          platformDigest = m.digest;
          break;
        }
      }

      // Fall back to first non-attestation manifest
      if (!platformDigest) {
        for (const m of manifests) {
          if (m.platform?.os !== "unknown") {
            platformDigest = m.digest;
            break;
          }
        }
      }

      if (!platformDigest) return null;
      if (depth >= 3) return null;
      return resolveManifestToConfigDigest(baseUrl, repoPath, platformDigest, token, depth + 1);
    }

    // Single manifest - extract config digest
    return data.config?.digest ?? null;
  } catch {
    return null;
  }
}

/**
 * Fetch OCI labels from an image's config blob without pulling the image.
 */
export async function fetchOciLabels(
  ref: ImageRef
): Promise<Record<string, string>> {
  const repoPath = `${ref.namespace}/${ref.name}`;
  const token = await getRegistryToken(ref.registry, repoPath);
  if (!token) return {};

  const baseUrl = registryBaseUrl(ref.registry);
  const configDigest = await resolveManifestToConfigDigest(
    baseUrl,
    repoPath,
    ref.tag,
    token
  );
  if (!configDigest) return {};

  try {
    const resp = await fetch(`${baseUrl}/v2/${repoPath}/blobs/${configDigest}`, {
      headers: {
        Authorization: `Bearer ${token}`,
        Accept: "application/vnd.docker.container.image.v1+json",
      },
      redirect: "follow",
      signal: AbortSignal.timeout(10000),
    });
    if (!resp.ok) return {};
    const data = await resp.json();
    return data.config?.Labels ?? {};
  } catch {
    return {};
  }
}
